### PR TITLE
quincy: pybind/mgr/pg_autoscaler: Cut back osdmap.get_pools calls

### PR DESCRIPTION
**Pull request by kamoltat**
**Problem:**

we have been extravagantly using osdmap get calls
which has led to performance issue when `pg_num`
is around ~3000.
    
**Solution:**

Limit the use of osdmap get calls by calling it once
and use the data for both functions `maybe_adjust`
and `update_progress_events`.
 
 Fixes: https://tracker.ceph.com/issues/62165
 
Signed-off-by: Kamoltat <ksirivad@redhat.com>

**Before Fix:**
pg_autoscaler-reef-5s-scrape-1000-osds-32768pg

**After Fix:**
pg_autoscaler-reef-patched-5s-scrape-1000-osds-32768pg

Backport tracker: https://tracker.ceph.com/issues/62886

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by kamoltat** [comment]
jenkins test make check

**Comment by kamoltat** [comment]
jenkins test make check

**Comment by kamoltat** [comment]
jenkins test api

**Comment by kamoltat** [comment]
jenkins test make check

**Comment by kamoltat** [comment]
jenkins test api

**Comment by kamoltat** [comment]
@NitzanMordhai do you mind leaving a review/approval for this PR? just wanna move it to qa

File: src/pybind/mgr/pg_autoscaler/module.py [hunk 1]
```diff
@@ -321,9 +321,11 @@ def set_noautoscale(self) -> Tuple[int, str, str]:
     def serve(self) -> None:
         self.config_notify()
         while not self._shutdown.is_set():
-            if not self.has_noautoscale_flag():
-               self._maybe_adjust()
-               self._update_progress_events()
+            if not self.noautoscale:
```
Comment (NitzanMordhai) [inline src/pybind/mgr/pg_autoscaler/module.py:324]: @kamoltat please check that again, it looks like in main we do have noautoscale and in quincy we are using the has_noautoscale_flag

**Review by NitzanMordhai — CHANGES_REQUESTED** [review]
Please check my comment

File: src/pybind/mgr/pg_autoscaler/module.py (hunk 1 above)
Comment (kamoltat) [reply src/pybind/mgr/pg_autoscaler/module.py:324]: Great catch @NitzanMordhai I must have been asleep while resolving the merge conflict of this backport.

**Comment by kamoltat** [comment]
@NitzanMordhai Okay just made the change, please take a look one more time, thank you!

**Comment by kamoltat** [comment]
jenkins test make check

**Comment by kamoltat** [comment]
make check failed because: https://tracker.ceph.com/issues/64235#change-253720

this PR is addressing the change: https://github.com/ceph/ceph/pull/55363

**Review by NitzanMordhai — APPROVED** [review]
LGTM

**Comment by kamoltat** [comment]
jenkins test make check

**Comment by ljflores** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/QUINCY#httpstrackercephcomissues65015

**Pull request merged**